include/envoy/formatter/substitution_formatter.h

```diff
@@ -78,5 +79,47 @@ class FormatterProvider {
 
 using FormatterProviderPtr = std::unique_ptr<FormatterProvider>;
 
+/**
+ * Interface for command parser.
+ * CommandParser returns a FormatterProviderPtr after successfully parsing an access log format
+ * token, nullptr otherwise.
+ */
+class CommandParser {
+public:
+  virtual ~CommandParser() = default;
+
+  /**
+   * Return a FormatterProviderPtr if this command is parsed from the token.
+   * @param token the token to parse
+   * @param pos current position in the entire format string
+   * @param command_end_position position at the end of the command token
+   * @return FormattterProviderPtr substitution provider for the parsed command
+   */
+  virtual FormatterProviderPtr parse(const std::string& token, size_t pos,
+                                     int command_end_position) const PURE;
+};
+
+using CommandParserPtr = std::unique_ptr<CommandParser>;
+
+/**
+ * Implemented by each custom CommandParser and registered via Registry::registerFactory()
+ * or the convenience class RegisterFactory.
+ */
+class CommandParserFactory : public Config::TypedFactory {
+public:
+  ~CommandParserFactory() override = default;
+
+  /**
+   * Creates a particular CommandParser implementation.
+   *
+   * @param config supplies the configuration for the command parser.
+   * @return CommandParserPtr the CommandParser which will be used in
+   * SubstitutionFormatParser::parse() when evaluating an access log format string.
+   */
+  virtual CommandParserPtr createCommandParserFromProto(const Protobuf::Message& config) PURE;
```

Comment:
I think you want a message validation context here as well.

Reply:
You mean within `createCommandParserFromProto()` something like `ProtobufMessage::getStrictValidationVisitor()` should be used or should it be a param for  `createCommandParserFromProto()` itself?

Looking at this other extension point (https://github.com/envoyproxy/envoy/pull/10908), the validator isn't part of creator method that receives a config...

Reply:
Yeah, I was looking into this recently [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/rate_limit_descriptors/expr/config.cc#L64), and it seems that validator generally comes from the factory context. A strict validator might be OK for now, not sure what additional benefit the context validator provides.